### PR TITLE
Fix to get only backoff metrics of discover RPC

### PR DESCRIPTION
**Pull request by hlts2**
#### WHAT

I fixed code to get only backoff metrics of discover RPC.

#### WHY

There was a lot of backoff metrics in discover RPC before.
The current implementation included a backoff metrics that is not directly related to the RPC call.

https://github.com/vdaas/vald/blob/master/internal/client/v1/client/discoverer/discover.go#L221

### Environment:

- Go Version: 1.18.3
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.6

### Types of changes:

- [x] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/client/v1/client/discoverer/discover.go [hunk 1]
```diff
@@ -218,111 +218,139 @@ func (c *client) dnsDiscovery(ctx context.Context, ech chan<- error) (addrs []st
...
+	for _, addr := range connectedAddrs {
+		cur.Store(addr, struct{}{})
+	}
 
 	for _, old := range oldAddrs {
 		_, ok := cur.Load(old)
 		if !ok {
 			c.eg.Go(safety.RecoverFunc(func() error {
-				err = c.disconnect(ctx, old)
+				err := c.disconnect(ctx, old)
```
Comment (kpango) [inline internal/client/v1/client/discoverer/discover.go:344]: can you describe about this?
I thought we can reuse returning definition.

File: internal/client/v1/client/discoverer/discover.go [hunk 2]
```diff
@@ -218,111 +218,139 @@ func (c *client) dnsDiscovery(ctx context.Context, ech chan<- error) (addrs []st
...
+}
 
-	oldAddrs := c.GetAddrs(ctx)
-
-	c.addrs.Store(connected)
+func (c *client) disconnectOldAddrs(ctx context.Context, oldAddrs, connectedAddrs []string, ech chan<- error) (err error) {
+	if !c.autoconn {
+		return nil
+	}
+	var cur sync.Map
```
Comment (kpango) [inline internal/client/v1/client/discoverer/discover.go:335]: I realized that we don't need sync.Map and we can disconnect concurrently

File: internal/client/v1/client/discoverer/discover.go [hunk 3]
```diff
@@ -218,111 +218,139 @@ func (c *client) dnsDiscovery(ctx context.Context, ech chan<- error) (addrs []st
...
+	for i := 0; i < maxPodLen; i++ {
+		for _, node := range nodes.GetNodes() {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+				if node != nil && node.GetPods() != nil {
+					pods := node.GetPods().GetPods()
+					if i < len(pods) {
+						addr := net.JoinHostPort(pods[i].GetIp(), uint16(c.port))
```
Comment (kpango) [inline internal/client/v1/client/discoverer/discover.go:310]:
```suggestion
				if node != nil && node.GetPods() != nil && len(node.GetPods().GetPods()) > i {
					addr := net.JoinHostPort(node.GetPods().GetPods()[i].GetIp(), uint16(c.port))
```

**Review by kpango — CHANGES_REQUESTED** [review]

File: internal/client/v1/client/discoverer/discover.go (hunk 1 above)
Comment (hlts2) [reply internal/client/v1/client/discoverer/discover.go]: Thank you for your comment.
I have fixed: https://github.com/vdaas/vald/pull/1706/commits/45057d7b5c5bdb82b61c570e511f6d81bb475e54

**Review by kpango — APPROVED** [review]

**Comment by hlts2** [comment]
Thank you for your review! 🙏 
I will merge this PR:+1:

**Pull request merged**